### PR TITLE
Issue 731: refine LedgerEntry interface and implementation

### PR DESCRIPTION
**Pull request by zhaijack**
In LedgerEntry interface, getEntry() and getEntryBuffer() have completely different behaviours. getEntry() releases bytebuf automatically, while getEntryBuffer() returns the bytebuf (if you don't call getEntry, you are responsible for releasing the entry buffer.

In this change:
make getEntry() reenter-able; make LedgerEntry implement AutoCloseable; LedgerEntry is responsible for releasing the bytebuf it is holding.

Descriptions of the changes in this PR:
1.  add `close` and `duplicate` in `api.LedgerEntry`,
1.  LedgerEntryImpl implements `api.LedgerEntry`,
1.  client.LedgerEntry doesn't implement the new api. it wraps an `api.LedgerEntry`,  
1.  add `close` in api.LastConfirmedAndEntry."
1.  fix build and test errors.

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
> 
> - [x] Make sure the PR title is formatted like:
>     `<Issue # or BOOKKEEPER-#>: Description of pull request`
>     `e.g. Issue 123: Description ...`
>     `e.g. BOOKKEEPER-1234: Description ...`
> - [x] Make sure tests pass via `mvn clean apache-rat:check install findbugs:check`.
> - [x] Replace `<Issue # or BOOKKEEPER-#>` in the title with the actual Issue/JIRA number.
> 
> ---

### CODE REVIEW

**Comment by asfgit** [comment]
FAILURE
 
--none--

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCounter.java [hunk 1]
```diff
@@ -22,6 +22,7 @@
 package org.apache.bookkeeper.client;
 
 import java.util.Enumeration;
+import org.apache.bookkeeper.client.api.LedgerEntry;
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCounter.java]: this seems not necessary

**Comment by asfgit** [comment]
FAILURE
 
--none--

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCounter.java (hunk 1 above)
Comment (eolivelli) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCounter.java]: Sync Counter is not used by current code, it is here only for backward compatibility.
Did you run all rests, even compatibility?

Comment (eolivelli) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCounter.java]: Sync Counter is not used by current code, it is here only for backward compatibility.
Did you run all rests, even compatibility?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/impl/LedgerEntryImpl.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    @Override
+    public ByteBuf getEntryBuffer() {
+        return entryBuf;
+    }
+
+    @Override
+    public LedgerEntryImpl duplicate() {
+        LedgerEntryImpl other = RECYCLER.get();
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/impl/LedgerEntryImpl.java]: this can be simplified with 

```
return duplicate(this);
```

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/sasl/GSSAPIBookKeeperTest.java [hunk 3]
```diff
@@ -1,13 +1,39 @@
...
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /*
```
Comment (sijie) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/sasl/GSSAPIBookKeeperTest.java]: since you are touching this file, you should move the LICENSE header to the beginning of this file.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/sasl/MD5DigestBookKeeperTest.java [hunk 4]
```diff
@@ -1,9 +1,27 @@
...
+import org.apache.bookkeeper.conf.ClientConfiguration;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.proto.BookieServer;
+import org.apache.bookkeeper.test.BookKeeperClusterTestCase;
+import org.junit.AfterClass;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /*
```
Comment (sijie) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/sasl/MD5DigestBookKeeperTest.java]: same as above

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerEntry.java [hunk 5]
```diff
@@ -1,114 +0,0 @@
...
-
-import java.io.InputStream;
-import org.apache.bookkeeper.conf.ClientConfiguration;
-
-/**
- * Ledger entry. Its a simple tuple containing the ledger id, the entry-id, and
- * the entry content.
- *
- */
-public class LedgerEntry
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerEntry.java]: I think you still need to keep this class for BC issue.

you can extend it from LedgerEntryImpl and override the getEntry() and keep getEntryInputStream(). because getEntry and getEntryInputStream should stay same semantics as legacy semantic in this class.

```
public class LedgerEntry extends LedgerEntryImpl {

     

}
```

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCounter.java (hunk 1 above)
Comment (jiazhai) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCounter.java]: Thanks. will re-run the compatibility tests.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/sasl/GSSAPIBookKeeperTest.java (hunk 3 above)
Comment (jiazhai) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/sasl/GSSAPIBookKeeperTest.java]: thanks, will change it. Seems the comments of header has some problem, and make the ide not recognise it.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/sasl/MD5DigestBookKeeperTest.java (hunk 4 above)
Comment (jiazhai) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/sasl/MD5DigestBookKeeperTest.java]: Got it.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/impl/LedgerEntryImpl.java (hunk 2 above)
Comment (jiazhai) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/impl/LedgerEntryImpl.java]: Thanks. will change it.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCounter.java (hunk 1 above)
Comment (jiazhai) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCounter.java]: re-checked, the BC tests all passed.

**Comment by asfgit** [comment]
FAILURE
 
--none--

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerEntry.java [hunk 6]
```diff
@@ -23,41 +23,37 @@
...
 
 /**
  * Ledger entry. Its a simple tuple containing the ledger id, the entry-id, and
  * the entry content.
  *
+ * @Deprecated since 4.6.0, in favor of using {@link org.apache.bookkeeper.client.impl.LedgerEntryImpl}.
  */
-public class LedgerEntry
-    implements org.apache.bookkeeper.client.api.LedgerEntry {
+@Deprecated
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerEntry.java]: don't add `@Deprecated` at this release.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/AsyncCallback.java [hunk 7]
```diff
@@ -17,6 +17,7 @@
 package org.apache.bookkeeper.client;
 
 import java.util.Enumeration;
+import org.apache.bookkeeper.client.api.LedgerEntry;
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/AsyncCallback.java]: Oh I think there is a problem here - this is the legacy api, you can't change this to `api.LedgerEntry`. try remove this line and recompile, you will find it will not compile. I think `client.LedgerEntry` should extend `impl.LedgerEntryImpl`. The current approach for `client.LedgerEntry` doesn't really work.

Comment (jiazhai) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/AsyncCallback.java]: I C. need changes

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerEntry.java (hunk 6 above)
Comment (jiazhai) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerEntry.java]: get it.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java [hunk 8]
```diff
@@ -56,11 +56,9 @@
...
 
     /**
-     * Returns the content of the entry. This method can be called only once. While using v2 wire protocol this method
-     * will automatically release the internal ByteBuf.
+     * Returns the content of the entry.
      *
      * @return the content of the entry
-     * @throws IllegalStateException if this method is called twice
      */
     byte[] getEntry();
```
Comment (ivankelly) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java]: Returning a ByteBuf should be consistent whether using the V2 wire protocol or not. As it is now, the documentation for getEntryBuffer is super confusing.

Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java]: As in, are we returning ownership of the object? Or does the LedgerEntry own the buffer, and is this why it requires closing?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java [hunk 9]
```diff
@@ -32,7 +32,7 @@
  */
 @Public
 @Unstable
-public interface LedgerEntry {
+public interface LedgerEntry extends AutoCloseable {
```
Comment (ivankelly) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java]: Should probably be a different commit, but there is an inconsistency in the data types used in read and write.

WriteHandle is takes ByteBuf and ByteBuffer
ReadHandle/LedgerEntry returns ByteBuf and byte[]

This should be resolved.

**Comment by jiazhai** [comment]
@ivankelly , thanks for the comments, as @sijie pointed out, we may need keep old behavior for LedgerEntry, so may need to change this PR. will re-think and do the change later.

**Comment by ivankelly** [comment]
What is the old behaviour for LedgerEntry? The javadoc leads me to believe it's different depending on whether v2 protocol is used or not.

**Comment by sijie** [comment]
@ivankelly that is the 4.5 behavior of LedgerEntry. in the new api, we should change LedgerEntry to be a more consistent behavior.

**Comment by jiazhai** [comment]
updated this PR to keep old behavior of LedgerEntry; the tests that use client.LedgerEntry was un -touched.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java (hunk 9 above)
Comment (jiazhai) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java]: #750 opened for it.

Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java]: @ivankelly : a good point. it was inconsistent there because I asked @eolivelli to not use `byte[]`. but @eolivelli wants to have a `byte[]` array. that was probably what caused confusion in the initial commit. we can support all - `byte[]`, `ByteBuffer` and `ByteBuf`.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java (hunk 8 above)
Comment (jiazhai) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java]: In LedgerEntryImpl, LedgerEntry is reenterable,  closing is more for the ByteBuf release.

**Comment by sijie** [comment]
@ivankelly @eolivelli  can you guys review this one again? since you reviewed that before.

**Comment by jiazhai** [comment]
Oh, sorry, not noticed your comments @sijie , already did the merge. But if there is any comment, I could resolved it in later PRs.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java [hunk 10]
```diff
@@ -56,25 +55,31 @@
...
-     * Returns the content of the entry. This method can be called only once. While using v2 wire protocol this method
-     * will automatically release the internal ByteBuf.
+     * Returns the content of the entry.
      *
      * @return the content of the entry
-     * @throws IllegalStateException if this method is called twice
      */
     byte[] getEntry();
 
     /**
```
Comment (ivankelly) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java]: What happens to the ownership of the returned ByteBuf? Is the caller responsible for releasing it? @sijie @zhaijack

Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java]: @ivankelly - This call doesn't retain any reference of the underneath bytebuf. if a caller releases the entry by #close(), it will release the underneath bytebuf. If the caller want to use the bytebuf beyond the lifecycle of this entry, it has to retain the buf by itself and then be responsible for releasing its retained refcnt. I have improved the comment in my subsequent change at #755 . 

https://github.com/apache/bookkeeper/pull/755/files

Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerEntry.java]: ah, excellent, will check it out there.
